### DISCUSSION TITLE
Final test of Raffle/Lottery Smart Contract failing with error: [FAIL. Reason: nonexistent request] 🙆🏻‍♂️

### DISCUSSION THREAD

**Discussion by alfheimrShiven** [Cyfrin/foundry-full-course-cu#535] [Q&A] [answered]
_Disclaimer: I've made the question very descriptive, hence it might come to you as a LONG one, but if you read through you will find my observations in here, that will save you a lot of debugging effort on your end. So pls stick with it. Thanks! 🙃_

Okay, so I wrote the final test in `RaffleTest.t.sol` as below:
```
 function testFulfillRandomWordsPicksWinnerResetsStateAndSendMoney()
        public
        raffleEnteredAndTimePassed
    {
        // Arrange
        // 1. Add multiple players to the Raffle
        uint256 totalPlayers = 5;
        uint256 startingIndex = 1; // as one player has already entered using the raffleEnteredAndTimePassed() modifier

        for (uint256 i = startingIndex; i < totalPlayers; i++) {
            address player = address(uint160(i));
            hoax(player, STARTING_BALANCE);
            raffle.enterRaffle{value: activeNetworkConfig.entranceFee}();
        }

        uint256 prize = (activeNetworkConfig.entranceFee * totalPlayers);

        // Act
        vm.recordLogs();
        raffle.performUpkeep(""); // will emit the request ID of the requestRandomWords() req. to Chainlink VRF
        Vm.Log[] memory logs = vm.getRecordedLogs();
        bytes32 requestId = logs[1].topics[1]; // tapping into the RequestedRandomWords event that we emit in the performUpKeep()

        VRFCoordinatorV2Mock(activeNetworkConfig.vrfCoordinator)
            .fulfillRandomWords(uint256(requestId), address(raffle));

        // Assert
        assert(raffle.getRecentWinner() != address(0));
        assert(uint256(raffle.getRaffleState()) == 0);
        assert(raffle.getLastTimeStamp() < block.timestamp);
        assert(
            raffle.getRecentWinner().balance ==
                (STARTING_BALANCE - activeNetworkConfig.entranceFee + prize)
        );
    }
```

On running `forge test --match-test testFulfillRandomWordsPicksWinnerResetsStateAndSendMoney` I get the following error:
```
Encountered 1 failing test in test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: nonexistent request] testFulfillRandomWordsPicksWinnerResetsStateAndSendMoney() (gas: 289796)

Encountered a total of 1 failing tests, 0 tests succeeded
```

To debug I ran
```
forge test --match-test testFulfillRandomWordsPicksWinnerResetsStateAndSendMoney -vvvv
````

and here's what I got
```
[285529] RaffleTest::testFulfillRandomWordsPicksWinnerResetsStateAndSendMoney() 
    ├─ [0] VM::prank(player: [0x44E97aF4418b7a17AABD8090bEA0A471a366305C]) 
    │   └─ ← ()
    ├─ [47826] Raffle::enterRaffle{value: 100000000000000000}() 
    │   ├─ emit EnteredRaffle(PLAYER: player: [0x44E97aF4418b7a17AABD8090bEA0A471a366305C])
    │   └─ ← ()
    ├─ [0] VM::warp(32) 
    │   └─ ← ()
    ├─ [0] VM::roll(2) 
    │   └─ ← ()
    ├─ [0] VM::deal(0x0000000000000000000000000000000000000001, 10000000000000000000 [1e19]) 
    │   └─ ← ()
    ├─ [0] VM::prank(0x0000000000000000000000000000000000000001) 
    │   └─ ← ()
    ├─ [23926] Raffle::enterRaffle{value: 100000000000000000}() 
    │   ├─ emit EnteredRaffle(PLAYER: 0x0000000000000000000000000000000000000001)
    │   └─ ← ()
    ├─ [0] VM::deal(0x0000000000000000000000000000000000000002, 10000000000000000000 [1e19]) 
    │   └─ ← ()
    ├─ [0] VM::prank(0x0000000000000000000000000000000000000002) 
    │   └─ ← ()
    ├─ [23926] Raffle::enterRaffle{value: 100000000000000000}() 
    │   ├─ emit EnteredRaffle(PLAYER: 0x0000000000000000000000000000000000000002)
    │   └─ ← ()
    ├─ [0] VM::deal(0x0000000000000000000000000000000000000003, 10000000000000000000 [1e19]) 
    │   └─ ← ()
    ├─ [0] VM::prank(0x0000000000000000000000000000000000000003) 
    │   └─ ← ()
    ├─ [23926] Raffle::enterRaffle{value: 100000000000000000}() 
    │   ├─ emit EnteredRaffle(PLAYER: 0x0000000000000000000000000000000000000003)
    │   └─ ← ()
    ├─ [0] VM::deal(0x0000000000000000000000000000000000000004, 10000000000000000000 [1e19]) 
    │   └─ ← ()
    ├─ [0] VM::prank(0x0000000000000000000000000000000000000004) 
    │   └─ ← ()
    ├─ [23926] Raffle::enterRaffle{value: 100000000000000000}() 
    │   ├─ emit EnteredRaffle(PLAYER: 0x0000000000000000000000000000000000000004)
    │   └─ ← ()
    ├─ [0] VM::recordLogs() 
    │   └─ ← ()
    ├─ [71022] Raffle::performUpkeep(0x) 
    │   ├─ [43604] VRFCoordinatorV2Mock::requestRandomWords(0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, 1, 3, 500000 [5e5], 1) 
    │   │   ├─ emit RandomWordsRequested(keyHash: 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, requestId: 1, preSeed: 100, subId: 1, minimumRequestConfirmations: 3, callbackGasLimit: 500000 [5e5], numWords: 1, sender: Raffle: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1])
    │   │   └─ ← 1
    │   ├─ emit RequestedRandomWords(vrfRequestId: 1)
    │   └─ ← ()
    ├─ [0] VM::getRecordedLogs() 
    │   └─ ← [([0x63373d1c4696214b898952999c9aaec57dac1ee2723cec59bea6888f489a9772, 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, 0x0000000000000000000000000000000000000000000000000000000000000001, 0x00000000000000000000000062c20aa1e0272312bc100b4e23b4dc1ed96dd7d1], 0x000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000640000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000007a1200000000000000000000000000000000000000000000000000000000000000001), ([0xc3d0c9fbc1987665462eec0262eb7fc743a27d411fd3a0b9c20c1c64d99347d3, 0x0000000000000000000000000000000000000000000000000000000000000001], 0x)]
    ├─ [2847] VRFCoordinatorV2Mock::fulfillRandomWords(1, Raffle: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1]) 
    │   └─ ← "nonexistent request"
    └─ ← "nonexistent request"
```

As we can clearly understand the error is being thrown by `VRFCoordinatorV2Mock::fulfillRandomWords(1, Raffle: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1])l`. One more important thing to notice is the request ID being emitted after invoking the function `requestRandomWords(...)` and that my friend is logged as ` emit RequestedRandomWords(vrfRequestId: 1)` . Now this proves that the request ID is '1' and it's also being sent as a param to VRFCoordinatorV2Mock's `fulfillRandomWords(1, Raffle: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1])` which even though is getting the right request ID, throws the error `nonexistent request`. 💁🏽

To dig in further, I went into `VRFCoordinatorV2Mock.sol` and checked the functions we're calling:
1. `requestRandomWords(...)`
This function regenerates the `requestId=1` for `_subId=1` and adds it to the mapping `s_requests`. 
```s_requests[requestId] = Request({ subId: _subId, callbackGasLimit: _callbackGasLimit, numWords: _numWords });```
PS: Refer to source code.
It also emits `RandomWordsRequested` event as found in the logs:
```emit RandomWordsRequested(keyHash: 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, requestId: 1, preSeed: 100, subId: 1, minimumRequestConfirmations: 3, callbackGasLimit: 500000 [5e5], numWords: 1, sender: Raffle: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1])```

This log proves that there is a request with:
```
requestId: 1
subId: 1
```
which is added to the `s_request` mapping. 
Therefore if we `console.log(s_request[1].subId)` we'll get `1`.

Now lets check the final function:
2. ```fulfillRandomWords(...)```
This function has this check condition
```
if (s_requests[_requestId].subId == 0) {
            revert("nonexistent request");
        }
```
which throws the error that we're encountering. Now what I don't understand is how `if(s_requests[_requestId].subId == 0)` stands true when in the previous function `requestRandomWords(...)` it was proven that `s_requests[1].subId` returns `1` and NOT `0`. 

I know I wrote a lot, but I wanted to present my observations to help reduce your debugging time. Thanks!

**Comment by theirrationalone** [comment]
You forgot to do... vm.warp, and vm.roll Again before performUpkeep. 

Please put... 
vm.warp & vm.roll before performUpkeep. 

<br>
<br>

# Like This:

<br>

```solidity
function testFulfillRandomWordsPicksWinnerResetsStateAndSendMoney()
        public
        raffleEnteredAndTimePassed
    {
        // Arrange
        // 1. Add multiple players to the Raffle
        uint256 totalPlayers = 5;
        uint256 startingIndex = 1; // as one player has already entered using the raffleEnteredAndTimePassed() modifier

        for (uint256 i = startingIndex; i < totalPlayers; i++) {
            address player = address(uint160(i));
            hoax(player, STARTING_BALANCE);
            raffle.enterRaffle{value: activeNetworkConfig.entranceFee}();
        }

        uint256 prize = (activeNetworkConfig.entranceFee * totalPlayers);

        // ANSWER:
        vm.warp(block.timestamp + activeNetworkConfig.s_interval + 1);
        vm.roll(block.number + 1);

        // Act
        vm.recordLogs();
        raffle.performUpkeep(""); // will emit the request ID of the requestRandomWords() req. to Chainlink VRF
        Vm.Log[] memory logs = vm.getRecordedLogs();
        bytes32 requestId = logs[1].topics[1]; // tapping into the RequestedRandomWords event that we emit in the performUpKeep()

        VRFCoordinatorV2Mock(activeNetworkConfig.vrfCoordinator)
            .fulfillRandomWords(uint256(requestId), address(raffle));

        // Assert
        assert(raffle.getRecentWinner() != address(0));
        assert(uint256(raffle.getRaffleState()) == 0);
        assert(raffle.getLastTimeStamp() < block.timestamp);
        assert(
            raffle.getRecentWinner().balance ==
                (STARTING_BALANCE - activeNetworkConfig.entranceFee + prize)
        );
    }
```

## The raffleEnteredAndTimePassed would work only and will pass the time & mine the block only for that very first entry you performed in that modifier.

## And when new entrants entered into lottery again, the last timestamp updated into contract because of reentries of new players. So the interval of time again have to be passed relative to that very last timestamp that's why you have to perform warp & roll again after putting new players into the lottery.

Hope it helps 👆

Thank You! 🙏🙏🙏

**Reply by theirrationalone** [reply]
Note: The Winner can be anyone. You Should first log the result and check who the Winner is, then perform assertion on it.

```solidity
assert(raffle.getRecentWinner() != address(0)); // this will pass though, because Winner can't be 0x000000....
```

**Reply by theirrationalone** [reply]
Explanation of concept....

Let's Say for example:...

at Very first keep this in mind ->, When i deployed the contract "raffle" also called as lottery, It was 9:30 AM,

So we have a timestamp of 90:30AM <- called as lastTimestamp, and
We choose an interval of = 30 mintues,
and then i performed entry of new player into lottery using that modifier... "raffleEnteredAndTimePassed", Okay...

And We have tools/facilities to do things i.e.,

- warp for time-travelling.
- roll for mining blocks.

Thanks to foundry again.

As you know, The modifier "raffleEnteredAndTimePassed" also performs warp & roll after putting new player into lottery/raffle,
So when i performed or put new player into lottery, warp & roll might also  happened,
And please ignore roll here,
So Warp affected the time and that lastTimestamp updated to 10:00 AM + 1.

Now we have...

Players = 1 <- into lottery/raffle.
lastTimestamp = 10:00 AM + 1, 

And now let's come to "testFulfillRandomWordsPicksWinnerResetsStateAndSendMoney" test function...

Here when i performed new entries of players into entry....

After putting all the desired numbers of players into lottery/raffle, I should do warp & roll again.... why???
**because VRFKeepers/VRFAutomation precisely "performUpkeep" function will again check if 30 mintues have been passed or not.** If not then revert or if passed then perform further work...

So that's why we have to perform warp & roll after that for loop and before that performUpkeep function call. According to our example this time after performing warp & roll, performUpkeep function will be executed at 10:30 AM + 5(appx). (i took 5 here which is estimated okay, we would also have consumed some time to put new players into raffle/lottery).

**Reply by theirrationalone** [reply]
To understand vm.roll [go here](https://github.com/Cyfrin/foundry-full-course-f23/discussions/468#discussioncomment-6637107)

**Reply by alfheimrShiven** [reply]
Hey @theirrationalone for the explaination, but the solution you're suggesting did not work. Also if it reverted due to that, the error message should have been `Raffle__UpkeepNotNeeded()`.
The reason is, even though we're pushing the block time by 30 mins in the modifier `raffleEnteredAndTimePassed` after making the first player enter the raffle, we're not updating the `s_lastTimeStamp` variable of Raffle.sol. which is the variable on which the `performUpKeep()` makes the interval check. `s_lastTimeStamp` is only updated on:
1. Contract deployment
2. FulfillRandomWords()
So performing vm.warp() once in the modifier is more than enough to push the block timestamp above the interval required to satisfy `performUpKeep()`.

In short:
Contract deployed at: 09:00. Therefore `s_lastTimeStamp= 09:00`
First user enters the raffle and performs vm.warp(): 09:00 + 30 mins + 1 = 10:01 `block.timestamp= 10:01`
`block.timestamp - s_lastTimeStamp > interval`
(10:01 - 09:00) > interval
Hence should work.

**Reply by theirrationalone** [reply]
Ohhhh yes, i forgot to explain this... 

```solidity
VRFCoordinatorV2Mock(activeNetworkConfig.vrfCoordinator) .fulfillRandomWords(uint256(requestId), address(raffle)); 
```

But @shikhar229169 mitigated your problem, Thanks to you for informing me and Thanks to @shikhar229169 to explain that... 

My bad 🤦‍♂sorry for not explaining that.

**Reply by theirrationalone** [reply]
And i also forgot if that happens then upkeepNotNeeded will be triggered, Sorry for bad context explanation.

**Reply by alfheimrShiven** [reply]
No issues man!

**Comment by shikhar229169** [comment, marked as answer]
The problem is in setUp function of Test file:

```solidity
function setUp() external {
        // Deploy the contract to be tested
        DeployRaffle deployRaffle = new DeployRaffle();
        raffle = deployRaffle.run();

        // getting access to active network variables through HelperConfig() script
        HelperConfig helperConfig = new HelperConfig();
        (
            activeNetworkConfig.entranceFee,
            activeNetworkConfig.interval,
            activeNetworkConfig.vrfCoordinator,
            activeNetworkConfig.subscriptionId,
            activeNetworkConfig.gasLane,
            activeNetworkConfig.callbackGasLimit,
            activeNetworkConfig.link
        ) = helperConfig.activeNetworkConfig();

        vm.deal(PLAYER, STARTING_BALANCE);
    }
```

You are deploying a new Helper Config contract, and as a result of which a new coordinator contract is deployed which is not the one by which you created subscription.

So instead of deploying new helper config you need to get the helper config from the deploy scripts, 
So, you just need to return the Helper Config contract from the run function of deploy scripts (the similar way we are returning the Raffle contract), and can read the activeNetworkConfig from there.

Also, after fixing this there is one more error in your assert statement.
```solidity
assert(raffle.getLastTimeStamp() < block.timestamp);
```
Figure it yourself what the error is in the assert.

**Reply by alfheimrShiven** [reply]
Thanks for pointing both things out @shikhar229169 . 😅 This helped!
Also to answer your challenge, `block.timestamp` will be equal to` raffle.getLastTimeStamp()` as the `s_lastTimeStamp` is being updated to block.timestamp() at the end of `fulfillRandomWords()`!
